flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -177,32 +187,53 @@ private boolean addBuffer(BufferConsumer bufferConsumer) {
 			buffers.add(bufferConsumer);
 			return false;
 		}
-		checkState(inflightBufferSnapshot.isEmpty(), "Supporting only one concurrent checkpoint in unaligned " +
-			"checkpoints");
 
 		buffers.addPriorityElement(bufferConsumer);
 		final int numPriorityElements = buffers.getNumPriorityElements();
 
-		// Meanwhile prepare the collection of in-flight buffers which would be fetched in the next step later.
-		final Iterator<BufferConsumer> iterator = buffers.iterator();
-		Iterators.advance(iterator, numPriorityElements);
-		while (iterator.hasNext()) {
-			BufferConsumer buffer = iterator.next();
-
-			if (buffer.isBuffer()) {
-				try (BufferConsumer bc = buffer.copy()) {
-					inflightBufferSnapshot.add(bc.build());
+		CheckpointBarrier barrier = parseCheckpointBarrier(bufferConsumer);
+		if (barrier != null) {
+			checkState(
+				barrier.getCheckpointOptions().isUnalignedCheckpoint(),
+				"Only unaligned checkpoints should be priority events");
+			final Iterator<BufferConsumer> iterator = buffers.iterator();
+			Iterators.advance(iterator, numPriorityElements);
+			List<Buffer> inflightBuffers = new ArrayList<>();
+			while (iterator.hasNext()) {
+				BufferConsumer buffer = iterator.next();
+
+				if (buffer.isBuffer()) {
+					try (BufferConsumer bc = buffer.copy()) {
+						inflightBuffers.add(bc.build());
+					}
 				}
 			}
+			if (!inflightBuffers.isEmpty()) {
+				channelStateWriter.addOutputData(
+					barrier.getId(),
+					subpartitionInfo,
+					ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
+					inflightBuffers.toArray(new Buffer[0]));
+			}
 		}
```

Comment:
nit: move to a separate `processCheckpointBarrier` method?

or maybe also one step further:
```
	private boolean addBuffer(BufferConsumer bufferConsumer) {
		assert Thread.holdsLock(buffers);	
		if (bufferConsumer.getDataType().hasPriority()) {
			processPriorityBuffer(bufferConsumer);
		}
		buffers.add(bufferConsumer);
		return false;
	}
```

to highlight that on the hot path, (without priority) we are skipping this whole code?